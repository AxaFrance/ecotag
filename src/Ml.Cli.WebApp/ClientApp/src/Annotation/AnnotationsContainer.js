import React, {useEffect, useState} from "react";
import AnnotationSwitch from "./AnnotationSwitch";
import AnnotationsToolbar from "./AnnotationsToolbar";
import './AnnotationsContainer.scss';
import {useHistory} from "react-router";
import {useMutation} from "react-query";
import {fetchGetData, fetchPostJson, utf8_to_b64} from "../FetchHelper";
import {toast} from "react-toastify";

const selectItemById = (annotationState, id) => {
    if(id === "end")
        return null;
    if(id === "empty")
        return null;
    return annotationState.items.find(x => x.id === id);
};


const fetchImages = async data => {
    if (data.status === 200) {
        const hardDriveLocations = await data.json();
        return hardDriveLocations.map(element => {
            return `api/files/${utf8_to_b64(element)}`;
        });
    } else {
        return [];
    }
};

const getImages = (fetchFunction) => async (item) => {
    const params = "fileName=" + item.fileName + "&stringsMatcher="+ item.frontDefaultStringsMatcher + "&directory=" + item.imageDirectory
    const fetchResult = await fetchGetData(fetchFunction)("api/datasets/"+utf8_to_b64(params));
    return fetchImages(fetchResult);
};

const sendConfirmationMessage = (isSuccess) => {
    const message = isSuccess ? "Annotation saved" : "Impossible to save annotation";
    const type = isSuccess ? "success" : "error";
    toast(message, {
        position: "top-left",
        autoClose: 800,
        hideProgressBar: true,
        closeOnClick: true,
        pauseOnHover: false,
        draggable: true,
        progress: undefined,
        type: type
    });
}

const AnnotationsContainer = ({state, id, url, dataset, fetchFunction}) => {
    const history = useHistory();
    
    const item = selectItemById(state, id);
    const itemNumber= state.items.indexOf(item);
    
    const onSubmit = async (annotation) => {
        const annotationObject = {
            datasetLocation: state.datasetLocation,
            annotationType: state.annotationType,
            fileName: item.fileName,
            annotation: annotation
        };
        await mutationDataset.mutateAsync(annotationObject, {
            onSuccess: () => {
                sendConfirmationMessage(true);
                onNext();
            },
            onError: () => sendConfirmationMessage(false)
        });
    }
    
    const onPrevious = () => {
        if(id === "end"){
            history.push(`${url}/${dataset}/${state.items[state.items.length - 1].id}`);
        } else{
            history.push(`${url}/${dataset}/${state.items[itemNumber - 1].id}`);
        }
    };
    
    const onNext = () => {
        if((itemNumber + 1) < state.items.length){
            history.push(`${url}/${dataset}/${state.items[itemNumber + 1].id}`);
        } else {
            history.push(`${url}/${dataset}/end`);
        }
    };

    const isPreviousDisabled = itemNumber === 0;
    const isNextDisabled = item == null;

    const [localState, setState] = useState({
        filePrimaryUrl: "",
    });
    console.log(item);

    const mutationDataset = useMutation(newData => fetchPostJson(fetchFunction)("/api/annotations", newData));

    const getUrls = async () => {
        const newUrls = await getImages(fetchFunction)(item);
        const newFileUrl = newUrls != null ? newUrls[0] : "";
        return {fileUrls: newUrls, filePrimaryUrl: newFileUrl};
    };

    const getBoundingBoxes = () => {
        let boundingBoxes;
        try{
            const annotationsArray = JSON.parse(item.annotations);
            const lastAnnotation = annotationsArray[annotationsArray.length - 1];
            const labels = lastAnnotation.labels;
            boundingBoxes = labels.boundingBoxes;
        }
        catch (ex){
            boundingBoxes = [];
        }
        return boundingBoxes;
    }

    useEffect(() => {
        setState({
            filePrimaryUrl: "",
        })
        if(!item || !item.imageDirectory){
            return;
        }
        getUrls().then(urls => {
                setState({
                    filePrimaryUrl: urls.filePrimaryUrl,
            })
        });
    }, [id]);

    if(state.items.length <= 0){
        return <h2 className="error-message">The annotation file is empty.</h2>
    }

    return <>
        <AnnotationsToolbar
            onPrevious={onPrevious}
            isPreviousDisabled={isPreviousDisabled}
            onNext={onNext}
            text={item? item.fileName: ""}
            isNextDisabled={isNextDisabled}
        />
        {isNextDisabled ? (
            <h3 className="annotation__end-message">Thank you, all files from this dataset have been annotated.</h3>
        ) : (
            <AnnotationSwitch
                url={localState.filePrimaryUrl}
                annotationType={state.annotationType}
                labels={state.configuration}
                expectedOutput={getBoundingBoxes()}
                onSubmit={onSubmit}
            />
        )}
    </>;
};

export default React.memo(AnnotationsContainer);
