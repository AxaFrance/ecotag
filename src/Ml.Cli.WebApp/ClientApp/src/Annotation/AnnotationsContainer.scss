.end-message{
  color: green;
  text-align: center;
}