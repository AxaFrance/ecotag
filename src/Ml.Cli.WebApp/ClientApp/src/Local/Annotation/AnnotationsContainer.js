import React, {useEffect, useState} from "react";

import {useHistory} from "react-router";
import {useMutation} from "react-query";
import {fetchGetData, fetchPostJson, utf8_to_b64} from "../FetchHelper";
import {toast} from "react-toastify";
import Annotations from "../../Toolkit/Annotations/Annotations";

const selectItemById = (annotationState, id) => {
    if(id === "end")
        return null;
    if(id === "empty")
        return null;
    return annotationState.items.find(x => x.id === id);
};


const fetchImages = async data => {
    if (data.status === 200) {
        const hardDriveLocations = await data.json();
        return hardDriveLocations.map(element => {
            return `api/local/files/${utf8_to_b64(element)}`;
        });
    } else {
        return [];
    }
};

const getImages = (fetch) => async (item) => {
    const params = "fileName=" + item.fileName + "&stringsMatcher="+ item.frontDefaultStringsMatcher + "&directory=" + item.imageDirectory
    const fetchResult = await fetchGetData(fetch)("api/local/datasets/"+utf8_to_b64(params));
    return fetchImages(fetchResult);
};

const sendConfirmationMessage = (isSuccess) => {
    const message = isSuccess ? "Annotation saved" : "Impossible to save annotation";
    const type = isSuccess ? "success" : "error";
    toast(message, {
        position: "top-left",
        autoClose: 800,
        hideProgressBar: true,
        closeOnClick: true,
        pauseOnHover: false,
        draggable: true,
        progress: undefined,
        type: type
    });
}


const AnnotationsContainer = ({state, id, url, dataset, fetch}) => {
    const history = useHistory();
    
    const item = selectItemById(state, id);
    const itemNumber= state.items.indexOf(item);
    
    const onSubmit = async (annotation) => {
        const annotationObject = {
            datasetLocation: state.datasetLocation,
            annotationType: state.annotationType,
            fileName: item.fileName,
            annotation: annotation
        };
        await mutationDataset.mutateAsync(annotationObject, {
            onSuccess: () => {
                sendConfirmationMessage(true);
                onNext();
            },
            onError: () => sendConfirmationMessage(false)
        });
    }
    
    const onPrevious = () => {
        if(id === "end"){
            history.push(`${url}/${dataset}/${state.items[state.items.length - 1].id}`);
        } else{
            history.push(`${url}/${dataset}/${state.items[itemNumber - 1].id}`);
        }
    };
    
    const onNext = () => {
        if((itemNumber + 1) < state.items.length){
            history.push(`${url}/${dataset}/${state.items[itemNumber + 1].id}`);
        } else {
            history.push(`${url}/${dataset}/end`);
        }
    };

    const isPreviousDisabled = itemNumber === 0;
    const isNextDisabled = item == null;

    const [localState, setState] = useState({
        filePrimaryUrl: "",
    });

    const mutationDataset = useMutation(newData => fetchPostJson(fetch)("/api/local/annotations", newData));

    const getUrls = async () => {
        const newUrls = await getImages(fetch)(item);
        const newFileUrl = newUrls != null ? newUrls[0] : "";
        return {fileUrls: newUrls, filePrimaryUrl: newFileUrl};
    };

    const getBoundingBoxes = () => {
        if(item && item.annotations){
            const annotationsArray = JSON.parse(item.annotations);
            const lastAnnotation = annotationsArray[annotationsArray.length - 1];
            return lastAnnotation.labels.boundingBoxes;
        }
        return null;
    }

    useEffect(() => {
        setState({
            filePrimaryUrl: "",
        })
        if(!item || !item.imageDirectory){
            return;
        }
        getUrls().then(urls => {
                setState({
                    filePrimaryUrl: urls.filePrimaryUrl,
            })
        });
    }, [id]);
    
    
    const toolbarText = item ? item.fileName: "";
    const annotationType = state.annotationType;
    const labels = state.configuration;
    const expectedOutput = getBoundingBoxes();

    return <Annotations
            onPrevious={onPrevious}
            isPreviousDisabled={isPreviousDisabled}
            onNext={onNext}
            toolbarText={toolbarText}
            isEmpty={state.items.length <= 0}
            isNextDisabled={isNextDisabled}
            url={localState.filePrimaryUrl}
            annotationType={annotationType}
            labels={labels}
            expectedOutput={expectedOutput}
            onSubmit={onSubmit}
        />;
};

export default React.memo(AnnotationsContainer);
