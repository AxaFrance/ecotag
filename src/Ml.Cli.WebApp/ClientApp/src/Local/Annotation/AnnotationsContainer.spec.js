import React from "react";
import {render, waitFor} from '@testing-library/react';
import AnnotationsContainer from "./AnnotationsContainer";
import {QueryClient, QueryClientProvider} from "react-query";

const fetch = () => {return {status: 400, ok: false}};

const client = new QueryClient();

const state = {
    datasetLocation: "datasetLocation",
    annotationType: "NamedEntityRecognition",
    items: [{
        filename: "filename.json",
        fileDirectory: "fileDirectory",
        imageDirectory: "",
        frontDefaultStringsMatcher: "",
        annotations: "",
        id: 1
    }],
    configuration: [{
        color: "#008194",
        id: "1",
        "name": "Firstname"
    }]
};

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

describe("Should check AnnotationsContainer component behaviour", () => {
    test("Should display annotation", async () => {
        const {container, asFragment} = render(<QueryClientProvider client={client}><AnnotationsContainer state={state} id={1} dataset="someDataset" url="someUrl" fetchFunction={fetch}/></QueryClientProvider>);
        
        expect(asFragment()).toMatchSnapshot();
        
        //necessary to force a longer waitFor
        await sleep(3000);
        await waitFor(() => expect(container.querySelector(".tokenAnnotation-container")).not.toBeNull());
        
        expect(asFragment()).toMatchSnapshot();
    })
})