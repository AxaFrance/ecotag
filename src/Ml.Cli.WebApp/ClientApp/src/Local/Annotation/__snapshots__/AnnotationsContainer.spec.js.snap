// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should check AnnotationsContainer component behaviour Should display annotation 1`] = `
<DocumentFragment>
  <div
    class="annotation__top-toolbar"
  >
    <button
      class="btn af-btn af-btn--hasiconLeft af-btn--disabled"
      disabled=""
    >
      <span
        class="af-btn__text"
      >
        Previous
      </span>
      <i
        class="glyphicon glyphicon-arrowthin-left"
      />
    </button>
    <span
      class="annotation__top-toolbar-text"
    />
    <button
      class="btn af-btn af-btn--hasiconRight"
    >
      <span
        class="af-btn__text"
      >
        Next
      </span>
      <i
        class="glyphicon glyphicon-arrowthin-right"
      />
    </button>
  </div>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;

exports[`Should check AnnotationsContainer component behaviour Should display annotation 2`] = `
<DocumentFragment>
  <div
    class="annotation__top-toolbar"
  >
    <button
      class="btn af-btn af-btn--hasiconLeft af-btn--disabled"
      disabled=""
    >
      <span
        class="af-btn__text"
      >
        Previous
      </span>
      <i
        class="glyphicon glyphicon-arrowthin-left"
      />
    </button>
    <span
      class="annotation__top-toolbar-text"
    />
    <button
      class="btn af-btn af-btn--hasiconRight"
    >
      <span
        class="af-btn__text"
      >
        Next
      </span>
      <i
        class="glyphicon glyphicon-arrowthin-right"
      />
    </button>
  </div>
  <div
    class="image-classifier"
  >
    <div
      class="image-classifier__buttons-container"
    >
      <div
        class="image-classifier__button-container"
      >
        <button
          class="btn af-btn"
          style="background-color: rgb(107, 149, 27); box-shadow: none;"
        >
          Firstname
        </button>
      </div>
    </div>
    <div
      class="image-classifier__image-container"
    >
      <img
        alt="Classifier image"
        class="image-classifier__image"
        src=""
      />
    </div>
  </div>
</DocumentFragment>
`;
